### ISSUE TITLE
Extra header support

### ISSUE THREAD

**Issue by MustafaJamal** [libgit2/libgit2#4249]
Library is working fine if we use UserNamePassword Credentials or Default Credentials with TFS 2017 BUT when we tried to use it with PAT/Access Tokens it failed to authenticate. So, we figured out that PAT/Access Token authentication requests are fulfilled successfully if we use extra header i.e.

`$ git -c http.extraheader='Authorization: Basic cGF0Ond0cTJ1M2tkaHJwc2U2ZzRreDQ0Z2p2NGg1cDJ1N3Y1N25pZnc2aGZ3cHFlenVob2ZoNWE=' ls-remote 'http://mjtfs2017:8080/tfs/DefaultCollection/_git/GITTest001/'`

We are not be able to find Extra Header support directly in LibGit2Sharp API, so is there any workaround possible ?

Thanks,
Mustafa

**Comment by shorelineai** [comment]
did you ever find the answer to this?

**Comment by MustafaJamal** [comment]
> 
> 
> did you ever find the answer to this?

Nope! never!

**Comment by tiennou** [comment]
You should be able to set up extra headers via [git_fetch_options](https://libgit2.org/libgit2/#HEAD/type/git_fetch_options). I have no idea if that's available in Libgit2Sharp though. `git_remote_connect` also can take custom headers.

**Comment by mariusin79** [comment]
@pks-t can you please give an explanation as to why this has been closed. @tiennou just mentions git fetch options, not how its possible to solve with libgit2. I am still looking for a libgit2 oauth 2.0 access token solution that works with Azure Devops

**Comment by ethomson** [comment]
It was closed because there's no bug here.  You can add additional headers per the documentation that was linked.  If you need assistance, please use [Stackoverflow](https://stackoverflow.com/questions/tagged/libgit2).

**Comment by manojbaishya** [comment]
Hi folks, I have almost a similar problem and posted a [detailed question on StackOverflow](https://stackoverflow.com/q/71126033/6483989), then happened to stumble into this issue.

> You should be able to set up extra headers via [git_fetch_options](https://libgit2.org/libgit2/#HEAD/type/git_fetch_options). I have no idea if that's available in Libgit2Sharp though. `git_remote_connect` also can take custom headers.

The `git_fetch_options` structure is in the plain C as well as the C# API (in the form of `FetchOptions` class), however it is only embedded as a field in the `CloneOptions` class, not `PushOptions`. This prevents the `Repository.Network.Push()` method from working with Azure DevOps repos with Personal Access Tokens. [Azure insists on passing the PAT on the Authentication header](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows#use-a-pat), so this puts us into a tricky situation. I am aware that this issue is talking about general headers support and `Clone` already has that, but it would be great if we can bring the `Push` to have the feature as well.

I looked around the [source code for `Push()`](https://github.com/libgit2/libgit2sharp/blob/1e6da83ab8c47058b5b4ef1e4e28e4732de44db0/LibGit2Sharp/Network.cs#L350)  and found that [in the `Proxy.git_remote_push()` call, we are instantiating a new `GitPushOptions` object in-place](https://github.com/libgit2/libgit2sharp/blob/1e6da83ab8c47058b5b4ef1e4e28e4732de44db0/LibGit2Sharp/Network.cs#L372).   [I hunted down this PR](https://github.com/libgit2/libgit2sharp/pull/1233) and to my surprise, [the `GitPushOptions` class actually contains a `CustomHeaders` field!](https://github.com/libgit2/libgit2sharp/blob/1e6da83ab8c47058b5b4ef1e4e28e4732de44db0/LibGit2Sharp/Core/GitPushOptions.cs#L12)

I believe that exposing the `GitPushOptions` class and allowing the consumer application to pass in custom headers into it will solve this issue, in the same vein as [what `Clone()` method is doing with `CloneOptions`](https://github.com/libgit2/libgit2sharp/blob/1e6da83ab8c47058b5b4ef1e4e28e4732de44db0/LibGit2Sharp/Repository.cs#L775).

I want to ask, does the `Push` method and `PushOptions` class need modifications to support this use case with Azure? If yes, then I will be very happy to submit a PR. I am hoping that we can author the code in the same manner as is present in the `Clone()` method, it will make the job easier, otherwise I might need some guidance. Please let me know your thoughts and inputs.

Regards, 
Manoj

**Comment by ethomson** [comment]
Yes, it probably does @manojbaishya, but you'd need to do that work in libgit2sharp, not here.

**Comment by manojbaishya** [comment]
Sorry for spamming again, but we have found a solution, it was simple: https://stackoverflow.com/a/71419019/6483989

**Comment by tapika** [comment]
> Sorry for spamming again, but we have found a solution, it was simple: https://stackoverflow.com/a/71419019/6483989

I tried this as well, but it does not seems to work.

Only walkaround I have found for using git command on azure is to execute git command manually, for example like this:

        /// <summary>
        /// Executes command
        /// </summary>
        /// <param name="token">authentication token</param>
        /// <param name="commandArgs">command line arguments</param>
        /// <param name="exitIfError">true if exit on error.</param>
        /// <returns></returns>
        bool GitCommand(BuildContext context, string token, string commandArgs, bool exitIfError = true)
        {
            var gitPath = context.Tools.Resolve("git.exe");
            if (gitPath == null)
                throw new Exception($"Unable to resolve git.exe tool.");

            string args = "";

            // Written according to https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows#use-a-pat
            if (!String.IsNullOrWhiteSpace(token))
            {
                string userPass = Convert.ToBase64String(Encoding.UTF8.GetBytes($":{token}"));
                args += $"-c http.extraHeader=\"Authorization: Basic {userPass}\" ";
            }

            args += commandArgs;

            StringBuilder sbError = new StringBuilder();
            var exitCode = context.StartProcess("git", new ProcessSettings { Arguments = args });

            if (exitIfError)
            {
                if (exitCode != 0)
                {
                    throw new Exception($"exit code: {exitCode} from command\n   git {args}");
                }
            }

            return exitCode != 0;
        }

And then:

* check if branch exists:

        Environment.CurrentDirectory = projectDir;
        if (!GitCommand(context, token, $"ls-remote --exit-code --heads origin {tobranch}", false))
                 throw new Exception($"Remote branch {tobranch} already exists.");

* delete remote branch:

        GitCommand(context, token, $"push origin --delete {tobranch}");

* git clone:

        void GitClone(BuildContext context, string url, string dir, string branchName, string token)
        {
            string args = "clone ";

            if (!string.IsNullOrEmpty(branchName))
            {
                args += $"-b {branchName} ";
            }
            args += $"{url} \"{dir}\"";

            GitCommand(context, token, args);
        }

* git push the changes:

    GitCommand(context, token, "push");


It would be nice if same operations could be performed with `LibGit2Sharp`.

Would it be possible to reopen this ticket ?

**Comment by ethomson** [comment]
> I tried this as well, but it does not seems to work.

What have you tried? The code you posted is the moral equivalent of using UsernamePasswordCredentials with an empty username and your PAT as the password, which libgit2  tests against Azure Repos and we expect to work.

**Comment by tapika** [comment]
So can Azure's PAT be used with `LibGit2Sharp` ? Can you post an example here on how it could be done ?
I guess git close is one use case, others which I've mentioned in here - like check if remote branch exists, etc...

**Comment by sebastianslutzky** [comment]
this ticket was closed without the sample code requested. I also cannot get  the PAT to work with Network.Push (tried this  https://stackoverflow.com/a/71419019/6483989 but it made no difference)
